Mathlib/Topology/ExtremallyDisconnected.lean

```diff
@@ -139,6 +136,132 @@ protected theorem CompactT2.Projective.extremallyDisconnected [CompactSpace X] [
     exact hx.1
 #align compact_t2.projective.extremally_disconnected CompactT2.Projective.extremallyDisconnected
 
+end
+
+section
+
+variable {A B C D : Type u} {E : Set D}
+  [TopologicalSpace A] [TopologicalSpace B] [TopologicalSpace C] [TopologicalSpace D]
+
+/-- Gleason's Lemma 2.4: a continuous surjection $\rho$ from a compact space $D$ to a Fréchet space
+$A$ restricts to a compact subset $E$ of $D$, such that $\rho$ maps $E$ onto $A$ and satisfies the
+"Zorn subset condition", where $\rho(E_0) \ne A$ for any proper closed subset $E_0 \subsetneq E$. -/
+lemma exists_compact_surjective_zorn_subset [T1Space A] [CompactSpace D] {π : D → A}
+    (π_cont : Continuous π) (π_surj : π.Surjective) : ∃ E : Set D, CompactSpace E ∧ π '' E = univ ∧
+    ∀ E₀ : Set E, E₀ ≠ univ → IsClosed E₀ → E.restrict π '' E₀ ≠ univ := by
+  let S : Set <| Set D := {E : Set D | IsClosed E ∧ π '' E = univ}
+  suffices ∀ (C : Set <| Set D) (_ : C ⊆ S) (_ : IsChain (· ⊆ ·) C), ∃ s ∈ S, ∀ c ∈ C, s ⊆ c by
+    rcases zorn_superset S this with ⟨E, ⟨E_closed, E_surj⟩, E_min⟩
+    refine ⟨E, isCompact_iff_compactSpace.mp E_closed.isCompact, E_surj, ?_⟩
+    intro E₀ E₀_min E₀_closed
+    contrapose! E₀_min
+    exact eq_univ_of_coe_eq <|
+      E_min E₀ ⟨E₀_closed.trans E_closed, image_coe_eq_restrict_image ▸ E₀_min⟩ coe_subset
+  intro C C_sub C_chain
+  refine ⟨iInter (fun c : C => c), ⟨isClosed_iInter fun ⟨_, h⟩ => (C_sub h).left, ?_⟩,
+    fun c hc _ h => mem_iInter.mp h ⟨c, hc⟩⟩
+  by_cases hC : Nonempty C
+  · refine eq_univ_of_forall fun a => inter_nonempty_iff_exists_left.mp ?_
+    refine iInter_inter (ι := C) (π ⁻¹' {a}) _ ▸
+      IsCompact.nonempty_iInter_of_directed_nonempty_compact_closed _
+      ?_ (fun c => ?_) (fun c => IsClosed.isCompact ?_) (fun c => ?_)
+    · replace C_chain : IsChain (· ⊇ ·) C := C_chain.symm
+      have : ∀ s t : Set D, s ⊇ t → _ ⊇ _ := fun _ _ => inter_subset_inter_left <| π ⁻¹' {a}
+      exact (directedOn_iff_directed.mp C_chain.directedOn).mono_comp (· ⊇ ·) this
+    · rw [← image_inter_nonempty_iff, (C_sub c.mem).right, univ_inter]
+      exact singleton_nonempty a
+    all_goals exact (C_sub c.mem).left.inter <| (T1Space.t1 a).preimage π_cont
+  · rw [@iInter_of_empty _ _ <| not_nonempty_iff.mp hC, image_univ_of_surjective π_surj]
+
+/-- Gleason's Lemma 2.1: if $\rho$ is a continuous surjection from a topological space $E$ to a
+topological space $A$ satisfying the "Zorn subset condition", then $\rho(G)$ is contained in the
+closure of $A \setminus \rho(E \setminus G)}$ for any open set $G$ of $E$. -/
+lemma image_subset_closure_compl_image_compl_of_isOpen {ρ : E → A} (ρ_cont : Continuous ρ)
```

Comment:
I realize you are eventually getting `E : Set D` in the final theorem, but here there is no reason to have `E` be anything other than a topological space, so you might as well shadow the variable (or else not declare `E : Set D` with `variable`, which is probably the best option). The same goes for `ExtremallyDisconnected.homeoCompactToT2` and `ExtremallyDisconnected.homeoCompactToT2_injective`